### PR TITLE
NAS-118906 / 22.12 / Nas 118906 - Refactor dialogs in PodShellComponent and PodLogsComponent

### PR DESCRIPTION
**Pull request by Raz-Dva**
Go to Apps -> Installed Applications 
Click on the three dots menu in the right corner of the application card. Choose Shell or Logs. 
In the dialog box choose the pod ang click Choose.
Once in Pod Logs or Pod Shell, click on Reconnect, select another pod and reconnect.

### CODE REVIEW

File: src/app/pages/applications/chart-releases/chart-releases.component.ts [hunk 1]
```diff
@@ -495,7 +493,7 @@ export class ChartReleasesComponent implements AfterViewInit, OnInit, OnDestroy
       width: '50vw',
       minWidth: '650px',
       maxWidth: '850px',
-      data: { appName: name, type: PodSelectDialogType.Shell },
+      data: { appName: name, title: 'Choose pod', type: PodSelectDialogType.Shell },
```
Comment (undsoft) [inline src/app/pages/applications/chart-releases/chart-releases.component.ts:496]: Needs to be translated.

File: src/app/pages/applications/dialogs/pod-select/pod-select-dialog.component.spec.ts [hunk 2]
```diff
@@ -97,5 +100,65 @@ describe('PodSelectDialogComponent', () => {
...
+    const objSpy = {
+      type: PodSelectDialogType.PodLogs,
+      customSubmit: (dialog: PodSelectDialogComponent) => mockCustomSubmit(dialog),
+      afterDialogInit: () => {},
+    };
+    const apps = [{ label: 'nextcloud-app1', value: 'nextcloud-app1' }, { label: 'nextcloud-app2', value: 'nextcloud-app2' }];
+    const pods = [{ label: 'nextcloud-pod1', value: 'nextcloud-pod1' }, { label: 'nextcloud-pod2', value: 'nextcloud-pod2' }];
+    const containers = [{ label: 'nextcloud-container1', value: 'nextcloud-container1' }, { label: 'nextcloud-container2', value: 'nextcloud-container2' }];
+    const fillForm = async (): Promise<any> => {
+      spectator.component.apps$ = of(apps);
```
Comment (undsoft) [inline src/app/pages/applications/dialogs/pod-select/pod-select-dialog.component.spec.ts:125]: I would not set these properties directly. They are not really meant to be interacted with outside of the component.
I would inject `MockWebsocketService` and re-mock necessary calls.

File: src/app/pages/applications/dialogs/pod-select/pod-select-dialog.component.ts [hunk 3]
```diff
@@ -81,27 +95,34 @@ export class PodSelectDialogComponent implements OnInit {
...
+        this.hasPool = false;
         this.loader.close();
       },
     });
   }
 
   onPodSelect(): void {
-    const pod = this.form.value.pods;
+    if (this.data.customSubmit) {
+      this.data.customSubmit(this);
```
Comment (undsoft) [inline src/app/pages/applications/dialogs/pod-select/pod-select-dialog.component.ts:111]: This goes back to our discussion about how entity-form ended up with so many options :)

How about extracting all submit functions to relevant components and just make dialog return something? Or at least have every consumer of dialog provide a submit function.

**Review by undsoft — CHANGES_REQUESTED** [review]

File: src/app/pages/applications/chart-releases/chart-releases.component.ts (hunk 1 above)
Comment (Raz-Dva) [reply src/app/pages/applications/chart-releases/chart-releases.component.ts]: Title translated in html template
```suggestion 
<h1 mat-dialog-title>{{ hasPool ? title : 'No Pools Found' | translate }}</h1>
```

File: src/app/pages/applications/dialogs/pod-select/pod-select-dialog.component.spec.ts (hunk 2 above)
Comment (Raz-Dva) [reply src/app/pages/applications/dialogs/pod-select/pod-select-dialog.component.spec.ts]: Dialog for PodLogsComponent extracted into separate component.

File: src/app/pages/applications/dialogs/pod-select/pod-select-dialog.component.ts (hunk 3 above)
Comment (Raz-Dva) [reply src/app/pages/applications/dialogs/pod-select/pod-select-dialog.component.ts]: All submit functions extracted to relevant components.

File: src/app/pages/applications/dialogs/pod-select-logs/pod-select-logs-dialog.component.html [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    <ix-select
+      formControlName="apps"
+      [label]="'Apps' | translate"
+      [required]="true"
+      [options]="apps$"
+    ></ix-select>
+    <ix-select
+      formControlName="pods"
+      [label]="'Pods' | translate"
+      [tooltip]="'Pods' | translate"
```
Comment (undsoft) [inline src/app/pages/applications/dialogs/pod-select-logs/pod-select-logs-dialog.component.html:14]: Unnecessary tooltip

File: src/app/pages/applications/dialogs/pod-select-logs/pod-select-logs-dialog.component.ts [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+import { AppLoaderService } from 'app/modules/loader/app-loader.service';
+import { ApplicationsService } from 'app/pages/applications/applications.service';
+import { WebSocketService } from 'app/services';
+
+export type LogsDialogFormValue = PodSelectLogsDialogComponent['form']['value'];
+
+@UntilDestroy()
+@Component({
+  templateUrl: './pod-select-logs-dialog.component.html',
+})
```
Comment (undsoft) [inline src/app/pages/applications/dialogs/pod-select-logs/pod-select-logs-dialog.component.ts:16]: Let's add ChangeDetection.OnPush to components

File: src/app/pages/applications/chart-releases/chart-releases.component.ts [hunk 6]
```diff
@@ -511,10 +516,25 @@ export class ChartReleasesComponent implements AfterViewInit, OnInit, OnDestroy
...
+        type: PodSelectDialogType.Logs,
+        customSubmit: (formValueDialog: PodDialogFormValue, appName: string) => {
+          this.logDialogSubmit(formValueDialog, appName);
+        },
+      },
     });
   }
 
+  shellDialogSubmit(formValue: PodDialogFormValue, appName: string): void {
+    this.router.navigate(new Array('/apps/1/shell/').concat([appName, formValue.pods, formValue.command]));
```
Comment (undsoft) [inline src/app/pages/applications/chart-releases/chart-releases.component.ts:531]: Or just `['/apps/1/shell', appName, formValue.pods, formValue.command]`

File: src/app/pages/applications/dialogs/pod-select-logs/pod-select-logs-dialog.component.spec.ts [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+    expect(valueLogsForm).toEqual({
+      Apps: 'chartRelease3',
+      Pods: 'pod3',
+      Containers: 'container31',
+      'Tail Lines': '500',
+    });
+  });
+
+  it('function customSubmit should be calls when Choose is pressed', async () => {
```
Comment (undsoft) [inline src/app/pages/applications/dialogs/pod-select-logs/pod-select-logs-dialog.component.spec.ts:90]: `customSubmit function should be called when...`

File: src/app/pages/applications/dialogs/pod-select-logs/pod-select-logs-dialog.component.ts [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        this.apps.push(chart.name);
+      });
+      this.fillForm();
+    });
+  }
+
+  fillForm(): void {
+    const appOptions = this.apps.map((app) => ({ label: app, value: app }));
+    this.apps$ = of(appOptions);
+    const app = appOptions.find((el) => el.value === this.selectedApp);
```
Comment (undsoft) [inline src/app/pages/applications/dialogs/pod-select-logs/pod-select-logs-dialog.component.ts:64]: el is a bad name.

File: src/app/pages/applications/dialogs/pod-select-logs/pod-select-logs-dialog.component.spec.ts [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+import { IxFormsModule } from 'app/modules/ix-forms/ix-forms.module';
+import { IxFormHarness } from 'app/modules/ix-forms/testing/ix-form.harness';
+import { AppLoaderModule } from 'app/modules/loader/app-loader.module';
+import { ApplicationsService } from 'app/pages/applications/applications.service';
+import {
+  PodSelectLogsDialogComponent,
+} from 'app/pages/applications/dialogs/pod-select-logs/pod-select-logs-dialog.component';
+
+describe('PodSelectLogsDialogComponent', () => {
+  let mockCustomSubmit: jest.Mock<any, any>;
```
Comment (undsoft) [inline src/app/pages/applications/dialogs/pod-select-logs/pod-select-logs-dialog.component.spec.ts:19]: Can be just `jest.Mock`.

File: src/app/pages/applications/dialogs/pod-select/pod-select-dialog.component.spec.ts [hunk 10]
```diff
@@ -16,6 +16,7 @@ describe('PodSelectDialogComponent', () => {
   let spectator: Spectator<PodSelectDialogComponent>;
   let loader: HarnessLoader;
   let form: IxFormHarness;
+  let mockCustomSubmit: jest.Mock<any, any>;
```
Comment (undsoft) [inline src/app/pages/applications/dialogs/pod-select/pod-select-dialog.component.spec.ts:19]: Just `jest.Mock`

File: src/app/pages/applications/pod-shell/pod-shell.component.ts [hunk 11]
```diff
@@ -75,63 +77,25 @@ export class PodShellComponent implements TerminalConfiguration {
...
-        name: 'command',
-        placeholder: helptext.podConsole.chooseCommand.placeholder,
-        value: this.command,
-      }],
-      saveButtonText: helptext.podConsole.choosePod.action,
-      customSubmit: (entityDialog) => this.onChooseShell(entityDialog),
-      afterInit: (entityDialog) => this.afterShellDialogInit(entityDialog),
-    };
+    this.mdDialog.open(PodSelectDialogComponent, {
+      width: '50vw',
```
Comment (undsoft) [inline src/app/pages/applications/pod-shell/pod-shell.component.ts:81]: width appears unnecessary

**Review by undsoft — APPROVED** [review]
I had weird 500 proxy errors when trying to download logs, even though it works on master and the code is pretty much the same.
Снимок экрана 2022-11-26 в 09 53 41

This could be my local problem though, so if you can download logs, we should be good.

In general new dialogs appear to be working correctly.

**Comment by Raz-Dva** [comment]
> I had weird 500 proxy errors ...

I have not seen such an error. There were no problems with downloading.

**Pull request merged**